### PR TITLE
Adjoint upgrades

### PR DESCRIPTION
**Pull request by Thenerdstation**
**Context:**
Now that operations are replacing templates, the assumption that adjoints can be defined as a single gate isn't always true. In these contexts, it may make the most sense to expand an operation first and then adjoint it.

Diego ran into this issue when trying to add `adjoint` to his Carry and Sum templates. 

I also went ahead and removed the `do_queue` argument on adjoint, making the signatures of transforms more aligned (see #1161)

**Description of the Change:**
Adjoint will now expand ops that don't have a native `adjoint` defined. 

**Benefits:**
Nearly all ops should now work out of the box with adjoint.

**Possible Drawbacks:**
Devices that can take advantage of certain non-decomposed ops may see significant overhead when forced to use the decomposed adjoint.

**Related GitHub Issues:**
#1161

### CODE REVIEW

File: pennylane/ops/qubit.py
Comment (Thenerdstation) [inline pennylane/ops/qubit.py:96]: I'm removing `do_queue` here since `qml.invisible` makes it redundant.

**Comment by Thenerdstation** [comment]
@josh146 I would also like to add diego on this PR

**Review by josh146 — CHANGES_REQUESTED** [review]
My favourite type of PR, where it is mostly deletions :)

I'm curious about the deletion of the abstract `adjoint` method though, it would feel useful to keep this in? Happy to approve once that is discussed though.

File: pennylane/operation.py
Comment (Thenerdstation) [reply pennylane/operation.py:605]: The reason I removed it is because it would fail my `hasattr` check. 

Although, now that I'm thinking of it, it might be better to keep this in and do a try/except instead. That way, we can either default to expand if we hit a `NotImplemented`, or raise the error if we hit anything else. I'll change it to that.

**Comment by Thenerdstation** [comment]
Dont you hate it when coverage goes down because you removed things lol

**Comment by josh146** [comment]
> Dont you hate it when coverage goes down because you removed things lol

I really don't get why this happens

File: pennylane/operation.py
Comment (josh146) [reply pennylane/operation.py:605]: Yeah that's nicer! I like it.

**Comment by Thenerdstation** [comment]
@DSGuala you shouldn't need to define adjoint for template-like ops after this is merged in.

File: pennylane/ops/qubit.py
Comment (DSGuala) [inline pennylane/ops/qubit.py:2147]: For my understanding: this was removed because QFT and other template-like operations will now default to expand?  As you mentioned in #1200#r610612234 ?

Comment (Thenerdstation) [reply pennylane/ops/qubit.py:2147]: Yep!

File: pennylane/transforms/adjoint.py
Comment (josh146) [inline pennylane/transforms/adjoint.py:98]: Love it

**Review by josh146 — APPROVED** [review]

**Pull request merged**
